### PR TITLE
add dmpnn encoder layer and suitable unit test

### PR DESCRIPTION
**Pull request by ARY2260**
# Pull Request Template
This PR adds the `DMPNNEncoderLayer` class and a suitable unit test.
### Why are we doing this?
The `DMPNNEncoderLayer` class will be the layer used to perform the message passing algorithm for the DMPNN model.
The role of the `DMPNNEncoderLayer` class is to generate molecule encodings in the following steps:
  - Message passing phase
  - Get new atom hidden states and readout phase
  - Concatenate the global features
 
## Type of change

Please check the option that is related to your PR.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [x] Documentations (modification for documents)

## Checklist

- [x] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
  - [x] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.32.0**)
  - [x] Run `mypy -p deepchem` and check no errors
  - [x] Run `flake8 <modified file> --count` and check no errors
  - [x] Run `python -m doctest <modified file>` and check no errors
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New unit tests pass locally with my changes
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

File: deepchem/models/torch_models/layers.py [hunk 1]
```diff
@@ -1219,6 +1219,342 @@ def inverse(self, y: Sequence) -> Tuple[torch.Tensor, torch.Tensor]:
...
 
+class DMPNNEncoderLayer(nn.Module):
+  """
+  Encoder layer for use in the Directed Message Passing Neural Network (D-MPNN) [1]_.
+
+  The role of the DMPNNEncoderLayer class is to generate molecule encodings in following steps:
+  - Message passing phase
+  - Get new atom hidden states and readout phase
+  - Concatenate the global features
+  ----------------------------------------------------------------------------------------------------------
```
Comment (rbharath) [inline deepchem/models/torch_models/layers.py:1230]: Probably better to just remove separator and have white space

File: deepchem/models/torch_models/layers.py [hunk 2]
```diff
@@ -1219,6 +1219,342 @@ def inverse(self, y: Sequence) -> Tuple[torch.Tensor, torch.Tensor]:
...
+  - a1, a2, a3, a4, a5 => atom features
+  - h1, h2, h3, h4, h5 => hidden states of atoms
+  - b12, b21, b23, b32, b24, b42, b15, b51 => bond features bonds
+  - (0)h12, (0)h21, (0)h23, (0)h32, (0)h24, (0)h42, (0)h15, (0)h51 => initial hidden states of bonds
+
+  The hidden state of every bond is a function of the concatenated feature vector which contains
+  concatenation of the 'features of initial atom of the bond' and 'bond features'.
+
+  Example: (0)h21 = func1(concat(a2, b21))  # here func1 is `self.W_i`
+  ------------------------------------------------------------------------------------------------------------
```
Comment (rbharath) [inline deepchem/models/torch_models/layers.py:1250]: Same here

**Comment by ARY2260** [comment]
The docstrings have been formatted for proper display in docs.
image
image
image
image
image
image

**Review by rbharath — APPROVED** [review]
LGTM

**Pull request merged**